internal/batch/batch_future_test.go

```diff
@@ -0,0 +1,297 @@
+				ScheduleToStartTimeout: time.Second * 10,
+				StartToCloseTimeout:    time.Second * 10,
+			})
+			return internal.ExecuteActivity(aCtx, batchActivity, i)
+		}
+	}
+
+	batchFuture, err := NewBatchFuture(ctx, input.Concurrency, factories)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]int, input.TotalSize)
+	err = batchFuture.Get(ctx, &result)
+	return result, err
+}
+
+func batchWorkflowUsingFutures(ctx internal.Context, input batchWorkflowInput) ([]int, error) {
+	factories := make([]func(ctx internal.Context) internal.Future, input.TotalSize)
+	for i := 0; i < input.TotalSize; i++ {
+		i := i
+		factories[i] = func(ctx internal.Context) internal.Future {
+			aCtx := internal.WithActivityOptions(ctx, internal.ActivityOptions{
+				ScheduleToStartTimeout: time.Second * 10,
+				StartToCloseTimeout:    time.Second * 10,
+			})
+			return internal.ExecuteActivity(aCtx, batchActivity, i)
+		}
+	}
+
+	batchFuture, err := NewBatchFuture(ctx, input.Concurrency, factories)
+	if err != nil {
+		return nil, err
+	}
+	result := make([]int, input.TotalSize)
+
+	for i, f := range batchFuture.GetFutures() {
+		err = f.Get(ctx, &result[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return result, err
+}
+
+func batchActivity(ctx context.Context, taskID int) (int, error) {
+	select {
+	case <-ctx.Done():
+		return taskID, fmt.Errorf("batch activity %d failed: %w", taskID, ctx.Err())
+	case <-time.After(time.Duration(rand.Int63n(100))*time.Millisecond + 900*time.Millisecond):
+		return taskID, nil
+	}
+}
+
+func Test_BatchWorkflow(t *testing.T) {
+	testSuite := &testsuite.WorkflowTestSuite{}
+	env := testSuite.NewTestWorkflowEnvironment()
+
+	env.RegisterWorkflow(batchWorkflow)
+	env.RegisterActivity(batchActivity)
+
+	totalSize := 100
+	concurrency := 20
+
+	startTime := time.Now()
+	env.ExecuteWorkflow(batchWorkflow, batchWorkflowInput{
+		Concurrency: concurrency,
+		TotalSize:   totalSize,
+	})
+
+	assert.Less(t, time.Since(startTime), time.Second*time.Duration(float64(totalSize)/float64(concurrency)))
+	assert.True(t, env.IsWorkflowCompleted())
+
+	assert.Nil(t, env.GetWorkflowError())
+	var result []int
+	assert.Nil(t, env.GetWorkflowResult(&result))
+	var expected []int
+	for i := 0; i < totalSize; i++ {
+		expected = append(expected, i)
+	}
+	assert.Equal(t, expected, result)
+}
+
+func Test_BatchWorkflow_Cancel(t *testing.T) {
+	testSuite := &testsuite.WorkflowTestSuite{}
+	env := testSuite.NewTestWorkflowEnvironment()
+	env.RegisterWorkflow(batchWorkflow)
+	env.RegisterActivity(batchActivity)
+
+	totalSize := 100
+	concurrency := 20
+
+	totalExpectedTime := time.Second * time.Duration(1+totalSize/concurrency)
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		env.ExecuteWorkflow(batchWorkflow, batchWorkflowInput{
+			Concurrency: concurrency,
+			TotalSize:   totalSize,
+		})
+	}()
+
+	time.Sleep(totalExpectedTime / 2)
+	env.CancelWorkflow()
```

Comment:
I'm a little surprised this works, but neat.

if you want to try fully simulating it, this is basically
```go
env.RegisterDelayedCallback(func() {
	env.CancelWorkflow()
}, totalExpectedTime / 2)
env.ExecuteWorkflow(...)
```

Reply:
My approach does have race condition. This is better and I've changed to this one.